### PR TITLE
Feature: adapter cte generation

### LINKED ISSUE #2609: Can't use CTEs in data test with dbt-sqlserver - Database Error
### Describe the bug
Using CTEs in data tests while using the dbt-sqlserver adapter raises SQL errors. This is due to the fact that data tests are wrapped around a `SELECT COUNT(*)` https://github.com/fishtown-analytics/dbt/blob/3af8a22d1319431115e7394f1235294613d36b35/core/dbt/task/test.py#L35 and SQL Server doesn't allow the CTEs inside other requests (it needs to start with WITH).

I has also been reported in the dbt-sqlserver repo [here](https://github.com/mikaelene/dbt-sqlserver/issues/26) but this seems to be actually due to the different behaviour of data tests compared to schema tests in dbt.

### Steps To Reproduce
- Configure dbt to run with dbt-sqlserver
- Create a test in the folder tests with a CTE (example: `WITH src AS (SELECT COUNT(*) AS cnt FROM {{ ref('my table') }})
) SELECT * FROM src` )
- Run dbt test --data
- A database error is raised (see logs below)

### Expected behavior
The test should run without any SQL error

### Screenshots and log output
From dbt.log
```
020-07-02 08:39:12,208505 (Thread-1): On test.qwdv.cte: select count(*) as errors from (

WITH src AS (
  <my_select_query>
)

SELECT * FROM src
) sbq
2020-07-02 08:39:12,213714 (Thread-1): Database error: ('42000', "[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Incorrect syntax near the keyword 'WITH'. (156) (SQLExecDirectW)")
2020-07-02 08:39:12,213930 (Thread-1): On test.qwdv.cte: ROLLBACK
```

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [ ] snowflake
- [X] other (specify: sqlserver)

**The output of `dbt --version`:**
```
installed version: 0.15.2
   latest version: 0.17.0
```

**The operating system you're using:**
Mac OS

**The output of `python --version`:**
Python 3.8.3

### Additional context
Is there any reason why the data tests are wrapped around a `SELECT COUNT(*)` and could the behavior be changed to the same as schema tests?

### LINKED ISSUE #2660: Remove double underscore in epheremal model generated CTE (__dbt__CTE__...)
### Describe the bug
In Exasol identifiers starting with double underscore are illegal - hence emepheral models fail when using dbt-exasol adapter.
 
### Steps To Reproduce
Configure you project to use dbt-exasol adapter (available on https://pypi.org/project/dbt-exasol/) and then create an ephemeral model. Once you use that model in a ref dbt generates a CTE named "__dbt__CTE__" which is not a valid identifier in exasol.

### Expected behavior
Ephemeral models should work in Exasol. Therefore we suggest to generate those CTE without the leading "__" characters.

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [ ] snowflake
- [x] other (specify: exasol)

**The output of `dbt --version`:**
```
All versions above 0.14.x
```

**The operating system you're using:**
ubuntu 18.04.1
**The output of `python --version`:**
Python 3.6.9

### Additional context
N/A

### PR DESCRIPTION
**Pull request by beckjake**
resolves #2660 (well, makes it feasible)
resolves #2609 

### Description
This PR makes it possible for the exasol plugin (or any others) to override their `Relation` class with one that has a custom `add_ephemeral_prefix` method. For the basic case, this should solve cte-naming issues.

The adapter class is further modified: It now provides a `get_compiler` method that returns a `Compiler` instance, so plugins can provide plugin-specific compilation changes.

In particular, it should now be possible to override how dbt creates ephemeral models. No promises, of course, but this should make it feasible for adapters that support subqueries but not CTEs to make their own "ephemeral" models. This is definitely left as an exercise to the plugin author, but overriding the handy `_insert_ctes` method should help. Note that `_insert_ctes` also does special work for making data tests work, so authors will need to re-implement that. Which is because...

This PR also makes data tests behave _much_ more like CTEs. Instead of performing a subquery, dbt inserts a custom CTE at the end of the CTE chain that includes the test body, and then the final select clause is a `select count(*) from` the CTE. Adapters that need to override CTE generation will need to do this as well, but I think that's ok. My understanding is that this will make #2609  work as it should out of the box.

Most adapters should not need to change anything. Any adapters that don't support CTEs but do support subqueries might break here. We should be able tof come up with a backwards-compatibility behavior here if we need to, and make only our adapters use this behavior.

Finally, while I was working on this I found that we were not properly excluding some things from the sandbox. This isn't a huge deal, but as a matter of hygiene we should be prefixing more attributes of objects in the context with `_` so jinja won't allow access.

### Checklist
 - [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [X] I have run this code in development and it appears to resolve the stated issue
 - [X] This PR includes tests, or tests are not required/relevant for this PR
 - [X] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: core/dbt/compilation.py [hunk 1]
```diff
@@ -186,6 +198,90 @@ def _get_compiled_model(
                 f'was not an ephemeral model: {cte_id}'
             )
 
+    def _inject_ctes_into_sql(self, sql: str, ctes: List[InjectedCTE]) -> str:
```
Comment (beckjake) [inline core/dbt/compilation.py:201]: This isn't new code, it moved from `core/dbt/contracts/graph/compiled.py`

File: core/dbt/compilation.py [hunk 2]
```diff
@@ -186,6 +198,90 @@ def _get_compiled_model(
...
+
+        token = sqlparse.sql.Token(
+            sqlparse.tokens.Keyword,
+            ", ".join(c.sql for c in ctes)
+        )
+        parsed.insert_after(with_stmt, token)
+
+        return str(parsed)
+
+    def _model_prepend_ctes(
```
Comment (beckjake) [inline core/dbt/compilation.py:262]: This isn't new code, it moved from `core/dbt/contracts/graph/compiled.py`, where it was previously a method on the `CompiledModel` type.

File: core/dbt/task/test.py [hunk 3]
```diff
@@ -41,10 +41,12 @@ def print_start_line(self):
         print_start_line(description, self.node_index, self.num_nodes)
 
     def execute_data_test(self, test: CompiledDataTestNode):
-        sql = (
-            f'select count(*) as errors from (\n{test.injected_sql}\n) sbq'
+        # sql = (
+        #     f'select count(*) as errors from (\n{test.injected_sql}\n) sbq'
+        # )
```
Comment (jtcohen6) [inline core/dbt/task/test.py:46]: Did you mean to leave this in?

**Comment by beckjake** [comment]
>That is, a user could not have CTEs within their data test definition, because the whole thing will be wrapped in another CTE. That's an extant problem for ephemeral models (fishtown-analytics/dbt-spark#11), too.

I don't particularly relish the thought of implementing this, but I think this PR makes it possible to solve that issue, since you can override the compiler to flatten CTEs. It does sound tricky, we might want to end up upstreaming that to dbt core if we do write it, for other plugins (to opt into: nesting them seems better if it's available).


Anyway, it should be pretty straightforward to override the data test behavior for those adapters.

File: core/dbt/task/test.py (hunk 3 above)
Comment (beckjake) [reply core/dbt/task/test.py:46]: no!

**Pull request merged**